clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2707,16 +2783,22 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
   static constexpr const bool VisitInsideSimpleContainers = false;
   SyclKernelDeclCreator(SemaSYCL &S, SourceLocation Loc, bool IsInline,
                         bool IsSIMDKernel, bool IsFreeFunction,
-                        FunctionDecl *SYCLKernel)
-      : SyclKernelFieldHandler(S), IsFreeFunction(IsFreeFunction),
-        KernelDecl(
-            createKernelDecl(S.getASTContext(), Loc, IsInline, IsSIMDKernel)),
+                        FunctionDecl *SYCLKernel, SYCLIntegrationHeader &H)
+      : SyclKernelFieldHandler(S), Header(H), IsFreeFunction(IsFreeFunction),
+        KernelDecl(IsFreeFunction
+                       ? createFreeFunctionDecl(S.getASTContext(), SYCLKernel,
+                                                Loc, IsInline)
+                       : createKernelDecl(S.getASTContext(), Loc, IsInline,
+                                          IsSIMDKernel)),
         FuncContext(SemaSYCLRef.SemaRef, KernelDecl) {
     S.addSyclOpenCLKernel(SYCLKernel, KernelDecl);
     for (const auto *IRAttr :
          SYCLKernel->specific_attrs<SYCLAddIRAttributesFunctionAttr>()) {
       KernelDecl->addAttr(IRAttr->clone(SemaSYCLRef.getASTContext()));
     }
+    if (IsFreeFunction)
+      FreeFunctionCDecl = createNewFunctionDecl(
```

Comment:
For the free function traits we need a declaration such as:
```
// Definition of _Z18__sycl_kernel_ff_1Piii as a free function kernel
void ff_1(int *ptr, int start, int end);
```

A templated free function prints as:
`template <> void ff_1(int *ptr, int start, int end);`

That's why a non-templated FunctionDecl is created, and then it can be printed in a form suitable to use in the int header definitions.

Reply:
Creating a `FunctionDecl` for this purpose is overkill. I would suggest creating your own print function to print what you need. It need not be a method of the visitor. It doesn't look like you need a visitor for your changes.

Reply:
Also, can you not have template parameters used as function arguments?

Reply:
OK, I've done the printing of the function declaration using inline code.